byterun/caml/backtrace_prim.h

```diff
@@ -36,22 +36,39 @@ struct caml_loc_info {
   int loc_lnum;
   int loc_startchr;
   int loc_endchr;
+  int loc_is_inlined;
 };
 
+/* When compiling with -g, backtrace slots have debug info associated.
+ * When a call is inlined in native mode, debuginfos form a linked list.
+ */
+typedef void * debuginfo;
+
 /* Check availability of debug information before extracting a trace.
  * Relevant for bytecode, always true for native code. */
 int caml_debug_info_available(void);
 
+/* Return debuginfo associated to a slot or NULL. */
+debuginfo caml_debuginfo_extract(backtrace_slot slot);
+
+/* In case of an inlined call return next debuginfo or NULL otherwise. */
+debuginfo caml_debuginfo_next(debuginfo dbg);
+
 /* Extract locations from backtrace_slot */
-void caml_extract_location_info(backtrace_slot pc,
-                                /*out*/ struct caml_loc_info * li);
-
-/* Expose a [backtrace_slot] as a OCaml value of type [raw_backtrace_slot].
- * The value returned should be an immediate and not an OCaml block, so that it
- * is safe to store using direct assignment and [Field], and not [Store_field] /
- * [caml_modify].  */
-value caml_val_raw_backtrace_slot(backtrace_slot pc);
-backtrace_slot caml_raw_backtrace_slot_val(value slot);
+void caml_debuginfo_location(debuginfo dbg, /*out*/ struct caml_loc_info * li);
+
+/* In order to prevent the GC from walking through the debug
+   information (which have no headers), we transform slots to 31/63 bits
+   ocaml integers by shifting them by 1 to the right. We do not lose
+   information as slots are aligned.
+
+   In particular, we do not need to use [caml_modify] when setting
+   an array element with such a value.
+
+   These functions might be removed in the future.
+ */
```

Comment:
I introduced these in a previous patch that defined the interface between backtrace.c and backtrace_prim.c

At this time, people were wondering if it was worth slowing down backtrace generation just for the sake of an abstraction internal to the compiler. So it was made clear in the comment that these "values" were immediates and caml_modify was not needed.

But since this is internal to the compiler and no real work as to be done for the conversion, I think macros are enough. Then I choose to change the naming from "backtrace_slot" to "debuginfo" as 1) slot is already overloaded 2) the real content of these values is the runtime representation of a Debuginfo.t, that they are put in a backtrace is not their "nature".

Both patches were intended to land in the same release, this one being a refinement of the previous one...
Turns out it got delayed quite a bit.